### PR TITLE
feat: Upload support cssVar

### PR DESCRIPTION
**Pull request by kiner-tang**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [X] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link

- #45618

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English | upload support css variable theme |
| 🇨🇳 Chinese | 上传支持 css 变量主题 |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [X] Doc is updated/provided or not needed
- [X] Demo is updated/provided or not needed
- [X] TypeScript definition is updated/provided or not needed
- [X] Changelog is provided or not needed

---

### <samp>🤖 Generated by Copilot at fde21ef</samp>

This pull request refactors the Upload component and its styles to use the new token system and the `@ant-design/cssinjs` library. It introduces a new `useCSSVar` hook in `./style/cssVar.ts` to register and apply CSS variables for the component. It also improves the consistency, readability, and maintainability of the CSS-in-JS code by using the `unit` and `calc` functions.

### <samp>🤖 Generated by Copilot at fde21ef</samp>

*  Add a new hook `useCSSVar` to register and apply CSS variables for the Upload component ([link](https://github.com/ant-design/ant-design/pull/45820/files?diff=unified&w=0#diff-561d8c96ea272bcbb7dfa3370e43215daa6d9dbccabf03aec149ecce487567d9R1-R4), [link](https://github.com/ant-design/ant-design/pull/45820/files?diff=unified&w=0#diff-a2e7152a9f97730a600f9bda0bf15029cce0148818a6ee8d2a037fafc101427dR21), [link](https://github.com/ant-design/ant-design/pull/45820/files?diff=unified&w=0#diff-a2e7152a9f97730a600f9bda0bf15029cce0148818a6ee8d2a037fafc101427dL366-R368), [link](https://github.com/ant-design/ant-design/pull/45820/files?diff=unified&w=0#diff-a2e7152a9f97730a600f9bda0bf15029cce0148818a6ee8d2a037fafc101427dL437-R439), [link](https://github.com/ant-design/ant-design/pull/45820/files?diff=unified&w=0#diff-a2e7152a9f97730a600f9bda0bf15029cce0148818a6ee8d2a037fafc101427dL468-R475))
*  Import and use the `unit` function from `@ant-design/cssinjs` to convert numeric values to CSS units and avoid potential calculation errors ([link](https://github.com/ant-design/ant-design/pull/45820/files?diff=unified&w=0#diff-3b4dbb97ee0c8cfbf0a7cae6e9aafaef93697d8552af9c0a29e5171b2bab7af6R1), [link](https://github.com/ant-design/ant-design/pull/45820/files?diff=unified&w=0#diff-3b4dbb97ee0c8cfbf0a7cae6e9aafaef93697d8552af9c0a29e5171b2bab7af6L15-R16), [link](https://github.com/ant-design/ant-design/pull/45820/files?diff=unified&w=0#diff-3b4dbb97ee0c8cfbf0a7cae6e9aafaef93697d8552af9c0a29e5171b2bab7af6L21-R22), [link](https://github.com/ant-design/ant-design/pull/45820/files?diff=unified&w=0#diff-3b4dbb97ee0c8cfbf0a7cae6e9aafaef93697d8552af9c0a29e5171b2bab7af6L53-R54), [link](https://github.com/ant-design/ant-design/pull/45820/files?diff=unified&w=0#diff-4292b1d3091a5dada24be824e64547feee97955fd1dce04d59a8866c35cdfccaR1), [link](https://github.com/ant-design/ant-design/pull/45820/files?diff=unified&w=0#diff-4292b1d3091a5dada24be824e64547feee97955fd1dce04d59a8866c35cdfccaL33-R34), [link](https://github.com/ant-design/ant-design/pull/45820/files?diff=unified&w=0#diff-f46faf24b775d47e5d87734c12b085fbbf31db15bda33d29b354b1dbc0c549c2L3-R10), [link](https://github.com/ant-design/ant-design/pull/45820/files?diff=unified&w=0#diff-f46faf24b775d47e5d87734c12b085fbbf31db15bda33d29b354b1dbc0c549c2L117-R121), [link](https://github.com/ant-design/ant-design/pull/45820/files?diff=unified&w=0#diff-f46faf24b775d47e5d87734c12b085fbbf31db15bda33d29b354b1dbc0c549c2L141-R146), [link](https://github.com/ant-design/ant-design/pull/45820/files?diff=unified&w=0#diff-f46faf24b775d47e5d87734c12b085fbbf31db15bda33d29b354b1dbc0c549c2L185-R189))
*  Import and use the `calc` function from the token object to perform arithmetic operations on the token values and ensure consistent units ([link](https://github.com/ant-design/ant-design/pull/45820/files?diff=unified&w=0#diff-647a2b5c69bacca02d0a7fc39f44e3749ec7e7583a83df0259315542f6f49668L3-R3), [link](https://github.com/ant-design/ant-design/pull/45820/files?diff=unified&w=0#diff-647a2b5c69bacca02d0a7fc39f44e3749ec7e7583a83df0259315542f6f49668L51-R71), [link](https://github.com/ant-design/ant-design/pull/45820/files?diff=unified&w=0#diff-4292b1d3091a5dada24be824e64547feee97955fd1dce04d59a8866c35cdfccaL6-R7), [link](https://github.com/ant-design/ant-design/pull/45820/files?diff=unified&w=0#diff-4292b1d3091a5dada24be824e64547feee97955fd1dce04d59a8866c35cdfccaL20-R21), [link](https://github.com/ant-design/ant-design/pull/45820/files?diff=unified&w=0#diff-4292b1d3091a5dada24be824e64547feee97955fd1dce04d59a8866c35cdfccaL80-R81), [link](https://github.com/ant-design/ant-design/pull/45820/files?diff=unified&w=0#diff-f46faf24b775d47e5d87734c12b085fbbf31db15bda33d29b354b1dbc0c549c2L3-R10), [link](https://github.com/ant-design/ant-design/pull/45820/files?diff=unified&w=0#diff-f46faf24b775d47e5d87734c12b085fbbf31db15bda33d29b354b1dbc0c549c2L22-R28), [link](https://github.com/ant-design/ant-design/pull/45820/files?diff=unified&w=0#diff-f46faf24b775d47e5d87734c12b085fbbf31db15bda33d29b354b1dbc0c549c2L35-R39), [link](https://github.com/ant-design/ant-design/pull/45820/files?diff=unified&w=0#diff-f46faf24b775d47e5d87734c12b085fbbf31db15bda33d29b354b1dbc0c549c2L54-R60), [link](https://github.com/ant-design/ant-design/pull/45820/files?diff=unified&w=0#diff-f46faf24b775d47e5d87734c12b085fbbf31db15bda33d29b354b1dbc0c549c2L93-R97), [link](https://github.com/ant-design/ant-design/pull/45820/files?diff=unified&w=0#diff-f46faf24b775d47e5d87734c12b085fbbf31db15bda33d29b354b1dbc0c549c2L157-R162), [link](https://github.com/ant-design/ant-design/pull/45820/files?diff=unified&w=0#diff-f46faf24b775d47e5d87734c12b085fbbf31db15bda33d29b354b1dbc0c549c2L222-R226), [link](https://github.com/ant-design/ant-design/pull/45820/files?diff=unified&w=0#diff-f46faf24b775d47e5d87734c12b085fbbf31db15bda33d29b354b1dbc0c549c2L237-R241))
*  Add a new property `listItemHeightSM` to the `ComponentToken` interface and calculate it based on the `fontSize` and `lineHeight` token values ([link](https://github.com/ant-design/ant-design/pull/45820/files?diff=unified&w=0#diff-647a2b5c69bacca02d0a7fc39f44e3749ec7e7583a83df0259315542f6f49668L17-R28))
*  Replace the second argument of the `genComponentStyleHook` function with the `prepareComponentToken` function to provide the default values for the Upload component ([link](https://github.com/ant-design/ant-design/pull/45820/files?diff=unified&w=0#diff-647a2b5c69bacca02d0a7fc39f44e3749ec7e7583a83df0259315542f6f49668L75-R85))
*  Use the `listItemHeightSM` property to set the height of the list item element in the Upload component ([link](https://github.com/ant-design/ant-design/pull/45820/files?diff=unified&w=0#diff-4292b1d3091a5dada24be824e64547feee97955fd1dce04d59a8866c35cdfccaL20-R21))
*  Use the `uploadThumbnailSize`, `uploadProgressOffset`, and `uploadPicCardSize` properties to set the size and position of the picture item and picture card elements in the Upload component ([link](https://github.com/ant-design/ant-design/pull/45820/files?diff=unified&w=0#diff-f46faf24b775d47e5d87734c12b085fbbf31db15bda33d29b354b1dbc0c549c2L22-R28), [link](https://github.com/ant-design/ant-design/pull/45820/files?diff=unified&w=0#diff-f46faf24b775d47e5d87734c12b085fbbf31db15bda33d29b354b1dbc0c549c2L35-R39), [link](https://github.com/ant-design/ant-design/pull/45820/files?diff=unified&w=0#diff-f46faf24b775d47e5d87734c12b085fbbf31db15bda33d29b354b1dbc0c549c2L54-R60), [link](https://github.com/ant-design/ant-design/pull/45820/files?diff=unified&w=0#diff-f46faf24b775d47e5d87734c12b085fbbf31db15bda33d29b354b1dbc0c549c2L157-R162), [link](https://github.com/ant-design/ant-design/pull/45820/files?diff=unified&w=0#diff-f46faf24b775d47e5d87734c12b085fbbf31db15bda33d29b354b1dbc0c549c2L222-R226), [link](https://github.com/ant-design/ant-design/pull/45820/files?diff=unified&w=0#diff-f46faf24b775d47e5d87734c12b085fbbf31db15bda33d29b354b1dbc0c549c2L237-R241))

### CODE REVIEW

File: components/upload/style/index.ts [hunk 1]
```diff
@@ -48,17 +54,21 @@ const genBaseStyle: GenerateStyle<UploadToken> = (token) => {
...
   (token) => {
-    const { fontSizeHeading3, fontSize, lineHeight, lineWidth, controlHeightLG } = token;
-    const listItemHeightSM = Math.round(fontSize * lineHeight);
+    const { fontSizeHeading3, listItemHeightSM, lineWidth, controlHeightLG, calc } = token;
 
     const uploadToken = mergeToken<UploadToken>(token, {
-      uploadThumbnailSize: fontSizeHeading3 * 2,
-      uploadProgressOffset: listItemHeightSM / 2 + lineWidth,
-      uploadPicCardSize: controlHeightLG * 2.55,
+      uploadThumbnailSize: calc(fontSizeHeading3).mul(2).equal(),
```
Comment (li-jia-nan) [inline components/upload/style/index.ts:69]: 闲夕大佬说 `mergeToken` 和 `prepareComponentToken` 两个地方都不用改，保持原样就好，只需要改消费的地方

Comment (li-jia-nan) [reply components/upload/style/index.ts:69]: image
image

Comment (MadCcc) [reply components/upload/style/index.ts:69]: mergeToken calc 还是要改的

File: components/upload/style/index.ts [hunk 2]
```diff
@@ -48,17 +54,21 @@ const genBaseStyle: GenerateStyle<UploadToken> = (token) => {
...
+      uploadProgressOffset: calc(calc(listItemHeightSM).div(2).equal()).add(lineWidth).equal(),
```
Comment (MadCcc) [inline components/upload/style/index.ts:70]:
```suggestion
      uploadProgressOffset: calc(calc(listItemHeightSM).div(2)).add(lineWidth).equal(),
```

**Comment by kiner-tang** [comment]
> image 这个看起来有点问题。其他的可能是用 github ci 机器跑的差异

这个我看看

**Comment by kiner-tang** [comment]
> > image 这个看起来有点问题。其他的可能是用 github ci 机器跑的差异
> 
> 这个我看看

done

**Comment by MadCcc** [comment]
我跑了一次 argos 还是有差异，应该是有问题

File: components/upload/style/index.ts [hunk 3]
```diff
@@ -48,17 +54,21 @@ const genBaseStyle: GenerateStyle<UploadToken> = (token) => {
   };
 };
 
+export const prepareComponentToken: GetDefaultToken<'Upload'> = (token) => ({
+  actionsColor: token.colorTextDescription,
+  listItemHeightSM: Math.round(token.fontSize * token.lineHeight),
```
Comment (MadCcc) [inline components/upload/style/index.ts:59]: 用 fontHeight

File: components/upload/style/picture.ts [hunk 4]
```diff
@@ -32,7 +37,7 @@ const genPictureStyle: GenerateStyle<UploadToken> = (token) => {
             ...textEllipsis,
             width: uploadThumbnailSize,
             height: uploadThumbnailSize,
-            lineHeight: `${uploadThumbnailSize + token.paddingSM}px`,
+            lineHeight: `${calc(uploadThumbnailSize).add(token.paddingSM).equal()}`,
```
Comment (MadCcc) [inline components/upload/style/picture.ts:40]:
```suggestion
            lineHeight: unit(calc(uploadThumbnailSize).add(token.paddingSM).equal()),
```

**Comment by kiner-tang** [comment]
> image 关掉 css var 后就看出来有问题了

是的，lineHeight 稍加一个单位导致的

**Review by MadCcc — APPROVED** [review]

**Pull request merged**